flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/calls/ScalarOperators.scala

```diff
@@ -656,38 +656,56 @@ object ScalarOperators {
       right: GeneratedExpression)
     : GeneratedExpression = {
 
-    val operator = if (plus) "+" else "-"
+    val op = if (plus) "+" else "-"
 
     (left.resultType, right.resultType) match {
       case (l: TimeIntervalTypeInfo[_], r: TimeIntervalTypeInfo[_]) if l == r =>
-        generateArithmeticOperator(operator, nullCheck, l, left, right)
+        generateArithmeticOperator(op, nullCheck, l, left, right)
 
       case (SqlTimeTypeInfo.DATE, TimeIntervalTypeInfo.INTERVAL_MILLIS) |
            (TimeIntervalTypeInfo.INTERVAL_MILLIS, SqlTimeTypeInfo.DATE) =>
         generateOperatorIfNotNull(nullCheck, SqlTimeTypeInfo.DATE, left, right) {
           if (isTimePoint(left.resultType)) {
-            (leftTerm, rightTerm) => s"$leftTerm + ((int) ($rightTerm / ${MILLIS_PER_DAY}L))"
+            (leftTerm, rightTerm) =>
+              s"$leftTerm $op ((int) ($rightTerm / ${MILLIS_PER_DAY}L))"
           } else {
-            (leftTerm, rightTerm) => s"((int) ($leftTerm / ${MILLIS_PER_DAY}L)) + $rightTerm"
+            (leftTerm, rightTerm) =>
+              s"((int) ($leftTerm / ${MILLIS_PER_DAY}L)) $op $rightTerm"
+          }
+        }
+
+      case (SqlTimeTypeInfo.DATE, TimeIntervalTypeInfo.INTERVAL_MONTHS) |
+           (TimeIntervalTypeInfo.INTERVAL_MONTHS, SqlTimeTypeInfo.DATE) =>
+        generateOperatorIfNotNull(nullCheck, SqlTimeTypeInfo.DATE, left, right) {
+          if (isTimePoint(left.resultType)) {
+            (leftTerm, rightTerm) =>
+              s"${qualifyMethod(BuiltInMethod.ADD_MONTHS.method)}($leftTerm, $op($rightTerm))"
+          } else {
+            (leftTerm, rightTerm) =>
+              s"${qualifyMethod(BuiltInMethod.ADD_MONTHS.method)}($rightTerm, $op($leftTerm))"
           }
         }
 
       case (SqlTimeTypeInfo.TIME, TimeIntervalTypeInfo.INTERVAL_MILLIS) |
            (TimeIntervalTypeInfo.INTERVAL_MILLIS, SqlTimeTypeInfo.TIME) =>
         generateOperatorIfNotNull(nullCheck, SqlTimeTypeInfo.TIME, left, right) {
           if (isTimePoint(left.resultType)) {
-            (leftTerm, rightTerm) => s"$leftTerm + ((int) ($rightTerm))"
+            (leftTerm, rightTerm) => s"$leftTerm $op ((int) ($rightTerm))"
           } else {
-            (leftTerm, rightTerm) => s"((int) ($leftTerm)) + $rightTerm"
+            (leftTerm, rightTerm) => s"((int) ($leftTerm)) $op $rightTerm"
           }
         }
 
       case (SqlTimeTypeInfo.TIMESTAMP, TimeIntervalTypeInfo.INTERVAL_MILLIS) =>
```

Comment:
Don't we need the inverse case `(TimeIntervalTypeInfo.INTERVAL_MILLIS, SqlTimeTypeInfo.TIMESTAMP)` as well?